### PR TITLE
fix(enricher): Streamline the pipeline to send flows from plugins to enricher

### PR DESCRIPTION
**Pull request by anubhabMajumdar**
## Description

Currently we have to many ways to extract out the events from plugins, and each plugin is responsible to init/check/send the data to enricher. Instead, setup a tracker module to "track" events generated and decide what to do with it. Currently, we simply forward to enricher.

## Testing done:

- Enabled in local context mode
- Can observe TCP, DNS and Drop advanced metrics
-  Basic mode works as expected
- Unit tests passing

### CODE REVIEW

File: pkg/track/events.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+var (
+	once sync.Once
+	t    *Track
+)
+
+func New() *Track {
+	once.Do(func() {
+		var e *enricher.Enricher
+		if e = enricher.Instance(); e == nil {
```
Comment (nddq) [inline pkg/track/events.go:35]: will cause the old error when we tried to compare a nil interface with nil (this will never be true)

Comment (anubhabMajumdar) [reply pkg/track/events.go:35]: Instance returns struct pointer, not interface - https://github.com/microsoft/retina/blob/f93df8d2ed54ca8fb77bf40d3dbb314c6275aea0/pkg/enricher/enricher.go#L61

Comment (anubhabMajumdar) [reply pkg/track/events.go:35]: Also, enricher should absolutely be a singleton for the entire lifecycle of Retina. It's a key piece of the control plane that needs the ringbuffer to be set and all events to be funneled to the same. The enricher also serves the enriched data to endpoint as metrics. All of this is predicated on a single enricher instance.

File: pkg/track/events.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import (
+	"context"
+	"sync"
+
+	v1 "github.com/cilium/cilium/pkg/hubble/api/v1"
+	"github.com/microsoft/retina/pkg/enricher"
+	"github.com/microsoft/retina/pkg/log"
+)
+
+type Track struct {
```
Comment (timraymond) [inline pkg/track/events.go:15]: Can you doc what a `Track` is intended to be?

File: pkg/track/events.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+	ctx             context.Context
```
Comment (timraymond) [inline pkg/track/events.go:16]: `context.Context` shouldn't be stored in a struct: see: https://pkg.go.dev/context

> Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it.

**Review by timraymond — CHANGES_REQUESTED** [review]

File: pkg/track/events.go (hunk 3 above)
Comment (anubhabMajumdar) [reply pkg/track/events.go:16]: Yes, `Start` take a context and lifecycle is tied to it. This variable is not actually needed, what I do need it the `cancelFunc` (in lieu of a separate done channel) to make sure I stop the track before I close the external channel.

File: pkg/track/events.go (hunk 2 above)
Comment (anubhabMajumdar) [reply pkg/track/events.go:15]: +1

File: Makefile [hunk 4]
```diff
@@ -338,12 +341,12 @@ test-image: ## build the retina container image for testing.
...
 			TAG=$(RETINA_PLATFORM_TAG)
 
-COVER_PKG ?= .
-
+.PHONY: test
 test: $(ENVTEST) # Run unit tests.
 	go build -o test-summary ./test/utsummary/main.go
 	CGO_ENABLED=0 KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use -p path)" go test -tags=unit -skip=TestE2E* -coverprofile=coverage.out -v -json ./... | ./test-summary --progress --verbose
 
+.PHONY: test-e2e
```
[the remark this answers is not in the document]
Comment (anubhabMajumdar) [reply Makefile:349]: +1

File: pkg/plugin/dns/dns_linux_test.go [hunk 5]
```diff
@@ -146,26 +137,31 @@ func TestRequestEventHandler(t *testing.T) {
...
-		utils.DNSType_QUERY, 0, event.DNSName, []string{event.QType}, 0, []string{},
-	)).Times(1)
-	d.enricher = mockEnricher
-
 	d.eventHandler(event)
 	after := value(c)
 	assert.Equal(t, after-before, float64(1))
+
+	// Test External channel.
+	em := EventMatched(utils.DNSType_QUERY, 0, "test.com", []string{"A"}, 0, []string{})
```
[the remark this answers is not in the document]
Comment (anubhabMajumdar) [reply pkg/plugin/dns/dns_linux_test.go:145]: +1

File: pkg/plugin/dns/dns_linux_test.go [hunk 6]
```diff
@@ -146,26 +137,31 @@ func TestRequestEventHandler(t *testing.T) {
...
+	select {
+	case ev := <-exCh:
+		assert.Assert(t, em.Matches(ev))
+	case <-time.After(3 * time.Second):
```
[the remark this answers is not in the document]
Comment (anubhabMajumdar) [reply pkg/plugin/dns/dns_linux_test.go:149]: +1

File: pkg/plugin/dns/dns_linux_test.go [hunk 7]
```diff
@@ -196,15 +192,18 @@ func TestResponseEventHandler(t *testing.T) {
...
 	d.eventHandler(event)
 	after := value(c)
 	assert.Equal(t, after-before, float64(1))
+
+	// Test External channel.
+	em := EventMatched(utils.DNSType_RESPONSE, 0, "test.com", []string{"A"}, 2, []string{"1.1.1.1", "2.2.2.2"})
+	select {
+	case ev := <-exCh:
+		assert.Assert(t, em.Matches(ev), "Expected event to match")
+	case <-time.After(3 * time.Second):
```
[the remark this answers is not in the document]
Comment (anubhabMajumdar) [reply pkg/plugin/dns/dns_linux_test.go:204]: +1

File: pkg/plugin/dropreason/dropreason_linux_test.go [hunk 8]
```diff
@@ -316,7 +310,8 @@ func TestDropReasonReadData_WithEmptyPerfArray(t *testing.T) {
 
 	// Start the drop reason routine in a goroutine
 	go func() {
-		dr.readEventArrayData()
+		err := dr.readEventArrayData()
+		assert.Nil(t, err, "Expected error but got nil")
```
[the remark this answers is not in the document]
Comment (anubhabMajumdar) [reply pkg/plugin/dropreason/dropreason_linux_test.go:314]: You are right, I have reworked them.

File: pkg/track/events.go [hunk 9]
```diff
@@ -0,0 +1,75 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+package track
```
Comment (anubhabMajumdar) [inline pkg/track/events.go:4]: We need to refactor the name to `TrackEvents`.

**Comment by timraymond** [comment]
@anubhabMajumdar Looks like stalebot killed this, but I see you manually closed this as well. What was the reason here?

**Pull request closed without merging**